### ISSUE TITLE
Accessibility - Label for Form Elements

### ISSUE THREAD

**Issue by tander08** [QualitativeDataRepository/dataverse#73]
## Resources: https://qdr.syr.edu/qdr-resources

- [ ] Search Website Bar under QDR Resources Header

**Comment by qqmyers** [comment]
Added aria-label="Search Website"

Although - I am confused as to why the Search Data box on the homepage doesn't raise an issue. As far as I can see, the only difference is that that one doesn't have a title, e.g. perhaps the title and placeholder together for the search box on the resources page causes confusion?

**Comment by adam3smith** [comment]
The search box is getting flagged by Firefox's built-in accessibility check, so I think that's just not getting picked up by Axe for whatever reason. We should add it there, too (can we dynamically do that so there are different labels depending on state?

**Comment by qqmyers** [comment]
The 'Search Data...' box I'm talking about is the one in the homepage, which doesn't change state. I see that FF raises the same issue for that,  the search icon in the header, and the search box that appears when you click that icon. It looks like FF is flagging the literal 'Form elements should have a visible label' whereas Axe says OK if there's an aria-label instead. We'd have to make a visible change in each of those three places, e.g. to add the word 'Search' somewhere nearby to fix it for FF as far as I can tell.

**Comment by adam3smith** [comment]
I'm pretty sure Axe is right overall. label is preferred where it's possible, but arial-label is acceptable where it isn't.

**Comment by adam3smith** [comment]
The Search data box on the homepage is now getting picked up by Axe (as is the box in the header...

**Comment by qqmyers** [comment]
I'm not seeing anything in Axe - is it in a different tool?

**Comment by adam3smith** [comment]
Yeah, sorry , this is in WAVE again (I thought I had seen it in AXE but apparently not). I think WAVE is correct here, though: I'm not seeing a label or Aria property. Here are the possible fixes it suggests:
 A properly associated label is:

-    a `<label>` element with a for attribute value that is equal to the id of a unique form control
-   a `<label>` element that surrounds the form control, does not surround any other form controls, and does not reference another element with its for attribute
-    a non-empty title attribute, or
-    a non-empty aria-labelledby attribute.

**Comment by adam3smith** [comment]
More issues with form labels: 
This is all testing in Firefox with WAVE (https://addons.mozilla.org/en-US/firefox/addon/wave-accessibility-tool/ )

On dataset creation initial page: https://data.stage-aws-new.qdr.org/dataset.xhtml?ownerId=1 

- [x] I'm seeing a number of empty Aria labels. All of those fields have marked labels, so should be fine, but this still seems not ideal? (or maybe doesn't matter?) This is for all standard fields, like title
- [x] Both WAVE and AXE complain about absent labels for the annotations upload fields
- [x] WAVE complains about some additional label-less fields, some of them hidden by CSS. I'm not 100% sure all of these need to be addressed, but let's take a look.
   - [x]  The (new?) Host Collection selector at the top
   - [x]  The Subject selector box
   - [x]  The Browse button for file uploads
   - [x]  The Terms of Access text box (hidden)
   - [x] The Provenance JSON Browse button (hidden)
   - [x] The Provenance Description (hidden)
   - [x] The deaccessioning textbox (I think) (hidden)

**Comment by qqmyers** [comment]
This page is down to two Wave errors - from empty links in the hidden close message elements generated for the main and provenance PrimeFaces fileupload widgets. I don't see any easy way to fix those - another upstream issue I think.

For the rest, most were just basic issues of missing/badly targetted labels. The empty aria-labels on many of the metadata elements was due to the way I was adding 'Additional Entry' aria-labels when you have multiple values (the main term name is the label for the first. I made a tweak to not add the aria-label attribute on the first value rather than leaving it with a blank value.
For the hidden computeBatch table - I added a facet for the last header column which gives it some text. I expect that is OK for IQSS as well but am not sure.
There are also a couple places where I tried to not render some hidden elements when they shouldn't be called. E.g. the there's an element telling you you can't publish a data project until you publish the parent and grandparent collections. If there is no grandparent, the link to it has no text, causing an error, but you should also never see that element if you don't have a grandparent collection. Similarly, the file info fragment usually displays info from a filemetadata object but it was also rendered (and I think always hidden) when there was no such object yet (with the null leading to missing text/labels, etc.). If these really can show under some circumstances, I'll need to look into some other fix (e.g. adding generic text/labels).

**Comment by adam3smith** [comment]
Checked on dev (doesn't look like it's on stage yet?) all looking great

**Comment by qqmyers** [comment]
yeah - sorry - updated the v5.9 branch but didn't get it deployed on stage before a call - doing it now.
